posts/2021-04-21-What I wish somebody told me when I was learning Haskell.md

```diff
@@ -0,0 +1,337 @@
+
+With Haskell it helps to differentiate between your application domain code, that you want to be as readable as possible, and the "glue code", that simply connects different things.
+
+Such glue are, for example, various "lift" functions that allow to compose different Monad types. Next time you stumble on something that looks like "lift", you can skip looking up what exactly it is - focus on the types around it and the functions that do something meaningful to your application - if compiler is happy and your code passes tests, understanding every bit of glue code is unnecessary.
+
+What helps is making reusable functions with short names that minimize the cognitive overhead.
+
+For example, one of the pieces of glue I found useful was this function:
+
+```haskell
+liftError :: (MonadIO m, MonadError e' m) => (e -> e') -> ExceptT e IO a -> m a
+liftError f a = liftIO (first f <$> runExceptT a) >>= liftEither
+```
+
+While it may look scary inside, like a computer, you don't need to understand how it works to be able to use it effectively - it is enough to understand by looking at its type that it "lifts" an action expressed as `ExceptT e IO` monad to a more generic monad `(MonadIO m, MonadError e' m)`, transforming the error type at the same time - it is useful when composing different libraries or modules each using their own error type.
+
+An equivalent code in `do` notation might be easier to parse for the beginners:
+
+```haskell
+liftError :: (MonadIO m, MonadError e' m) => (e -> e') -> ExceptT e IO a -> m a
+liftError f a = do
+  b <- liftIO $ runExceptT a
+  let c = case b of -- this case is Bifunctor `first` method used above
+        Left e -> Left (f e)
+        Right x -> Right x
+  liftEither c
+```
+
+But the point is that you do not need to understand either code to effectively use this function, and you will see that short functions are easier to maintain.
+
+## Embrace purity
+
+### Functions do nothing
+
+In a functional language a function does nothing - it simply defines a transformation between types - but it does not perform this transformation until it is... It is tempting to say "called", but the functions are not "called" in functional programming languages in the same way - they are evaluated and combined in bigger transformations, until the whole program is constructed.
+
+Only when the whole program is executed something happens - but you cannot simply break down program execution to the individual function calls - compiler can produce a more efficient machine code than if functions were individually called.
+
+Every function definition can be replaced with its code - there is no such thing as impure function in Haskell (and yes, that includes functions that operate on Monads, whether they use `do` notation or not, - they are also pure functions).
```

Comment:
I'm not an expert on English punctuation and as far as I know it's usually liberal, but I'm not sure there should be a comma before dash